index.js

```diff
@@ -877,67 +865,104 @@ class Watcher extends EventEmitter {
     this.opened = false
     this.closed = false
 
-    this.running = false
-
     this.latestDiff = 0
     this.range = range
+
+    this.current = null
+    this.previous = null
     this.stream = null
 
-    this._onappendBound = debounceify(this._onappend.bind(this))
+    this._lock = mutexify()
+    this._resolveOnChange = null
 
     this._opening = this._ready().catch(safetyCatch)
   }
 
   async _ready () {
     await this.bee.ready()
-    this.latestDiff = this.bee.version
+    this.current = this.bee.snapshot() // point from which to start watching
     this.opened = true
   }
 
-  async _onappend () {
-    this.running = true
+  [Symbol.asyncIterator] () {
+    return this
+  }
 
-    try {
-      await this._run()
-    } catch (err) {
-      if (!this.closed) this.emit('error', err)
-      this.destroy()
-    } finally {
-      this.running = false
-    }
+  _onappend () {
+    const resolve = this._resolveOnChange
+    this._resolveOnChange = null
+    if (resolve) resolve()
   }
 
-  async _run () {
-    if (this.closed) return
-    if (this.opened === false) await this._opening
+  async _waitForChanges () {
+    if (this.current.version < this.bee.version || this.closed) return
+
+    await new Promise(resolve => {
+      this._resolveOnChange = resolve
+    })
+  }
 
-    const snapshot = this.bee.snapshot()
-    this.stream = snapshot.createDiffStream(this.latestDiff, this.range)
+  async next () {
+    const release = await this._lock()
 
     try {
-      for await (const data of this.stream) { // eslint-disable-line
-        this.emit('change', snapshot.version, this.latestDiff)
-        break
+      if (this.closed) return { value: undefined, done: true }
+
+      if (!this.opened) await this._opening
+
+      while (true) {
+        await this._waitForChanges()
+        if (this.closed) return { value: undefined, done: true }
+
+        if (this.previous) await this.previous.close()
+        this.previous = this.current.snapshot()
+
+        if (this.current) await this.current.close()
+        this.current = this.bee.snapshot()
```

Comment:
Maybe overthinking but just trying to see if we should auto-destroy on error, but the lock complicates things because `destroy()` acquires a lock as well

Reply:
If one of these lines throws, doesn't it just bubble to the consumer and throw there, exiting the async iterator with the cleanup code being triggered automatically? (not sure, but that would be my intuition)

It's definitely possible for these lines to throw (e.g. if the user closed the `current` snapshot)

Reply:
Hm we don't have any automatic cleanup if it fails, besides the user destroying the watch or closing the bee if something fails